### PR TITLE
[macOS] The main thread in the WebContent process can block on startup

### PR DESCRIPTION
**Pull request by pvollan**
#### f8428becc59a0312a9513e03f0f9d14f26e938ab
<pre>
[macOS] The main thread in the WebContent process can block on startup
<a href="https://bugs.webkit.org/show_bug.cgi?id=264159">https://bugs.webkit.org/show_bug.cgi?id=264159</a>
<a href="https://rdar.apple.com/92981500">rdar://92981500</a>

Reviewed by NOBODY (OOPS!).

Calling [NSURL getResourceValue:forKey:error:] from [NSApplication _accessibilityInitialize] can block the main thread
for some time if NFS mounts become unavailable. This patch resolves this by temporarily swizzling
[NSURL getResourceValue:forKey:error:].

* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(NSURLgetResourceValue):
(WebKit::WebPage::platformInitializeAccessibility):
</pre>
https://github.com/WebKit/WebKit/commit/f8428becc59a0312a9513e03f0f9d14f26e938ab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24792 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3337 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26046 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26911 "Failed to checkout and rebase branch from PR 19950") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22761 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5016 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/774 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/26911 "Failed to checkout and rebase branch from PR 19950") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25037 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2383 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21407 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27494 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2090 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28490 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22613 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22683 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26304 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2027 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/332 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3318 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2478 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2379 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm [hunk 1]
```diff
@@ -113,18 +113,32 @@
...
     // For performance reasons, we should have received the LS database before initializing NSApplication.
     ASSERT(LaunchServicesDatabaseManager::singleton().hasReceivedLaunchServicesDatabase());
 
+    // Temporarily swizzle [NSURL getResourceValue:forKey:error:], since these calls can block the main thread. See rdar://92981500.
+    Method methodToPatch = class_getInstanceMethod([NSURL class], @selector(getResourceValue:forKey:error:));
+    auto originalNSURLgetResourceValueMethod = method_setImplementation(methodToPatch, (IMP)NSURLgetResourceValue);
+
     // Need to initialize accessibility for VoiceOver to work when the WebContent process is using NSRunLoop.
     // Currently, it is also needed to allocate and initialize an NSApplication object.
     [NSApplication _accessibilityInitialize];
```
Comment (brentfulgham) [inline Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:138]: Is it possible for this call to throw an exception? Should we wrap this in a `@try/@catch` to make sure we unswizzle?

I guess if we throw at that point, the whole WCP gets destroyed, so perhaps it's not worth worrying about.

**Review by brentfulgham — APPROVED** [review]
I think the change looks fine mechanically, but I'm not sure if replacing `getResourceValue` with a stub is safe here. Are we sure that `_accessibilityInitialize` really has no need to perform any URL reads? How would we know if we have introduced a new problem with this change?

I'd be interested to see if adding `NSLog` calls inside your stub showed any hits, and what those hits were for.

**Comment by aproskuryakov** [comment]
Wouldn't we hit the same stall the next time we need to read a file? It's not clear if this fixes any observable issues, or just pushes the problem to happen slightly later.

**Comment by pvollan** [comment]
> I think the change looks fine mechanically, but I'm not sure if replacing `getResourceValue` with a stub is safe here. Are we sure that `_accessibilityInitialize` really has no need to perform any URL reads? How would we know if we have introduced a new problem with this change?
> 
> I'd be interested to see if adding `NSLog` calls inside your stub showed any hits, and what those hits were for.

I verified in the debugger that `_accessibilityInitialize` has some calls to `getResourceValue`, and that the swizzled version returns the same values as the original one. However, more calls to `getResourceValue` could be added in the future, that we would potentially not return the same value for.

Thanks for reviewing!

**Comment by pvollan** [comment]
> Wouldn't we hit the same stall the next time we need to read a file? It's not clear if this fixes any observable issues, or just pushes the problem to happen slightly later.

That is a good point! I don't think there's any guarantee that we will not hit this later. So this may not be the right approach.

Thanks for reviewing!

**Comment by pvollan** [comment]
This issue has already been addressed.

**Pull request closed without merging**